Replace boost::unordered_{map,set} with std::unordered_{map,set}

sweemer opened the pull request:
Closes #1270 

- Replace `boost::unordered_map` and `boost::unordered_set` with the `std` equivalents.
- Deprecate unused (within QuantLib) public typedefs.
- Remove unnecessary warning suppression on Linux.
- Add harmless warning suppression on Windows.

lballabio commented on ql/patterns/observable.hpp:
```diff
@@ -89,18 +87,20 @@ namespace QuantLib {
     class Observer {
       public:
-        typedef boost::unordered_set<ext::shared_ptr<Observable> > set_type;
+        QL_DEPRECATED // TODO - make private and undeprecate
+        typedef std::unordered_set<ext::shared_ptr<Observable> > set_type;
+        QL_DEPRECATED // TODO - make private and undeprecate
         typedef set_type::iterator iterator;
```
If we deprecate, we have to leave the deprecated `boost::unordered_set` for the time being and change it to `std::` when it moves to private.

sweemer replied:
I've reverted back the publicly exposed typedefs to the boost version.

This pull rquest is still a draft and I may have a couple other changes to make in the next few days depending on the CI results.

lballabio commented on ql/timeseries.hpp:
```diff
@@ -110,15 +110,34 @@ namespace QuantLib {
         typedef typename Container::const_iterator const_iterator;
         typedef typename const_iterator::iterator_category iterator_category;
 
+#if __cplusplus < 201703L
+        template <class ...>
+        using void_t = void;
+#else
+        using std::void_t;
+#endif
+
+        template <class container, typename = void>
+        struct enable_reverse final : std::false_type {
+        };
+
+        template <class container>
+        struct enable_reverse<container, void_t<typename container::const_reverse_iterator>> final : std::true_type {
+        };
+
+        template <class container, typename = void>
+        struct reverse {
+        };
+
```
Before going into this, I would check if we really need this class to have a `Container` template argument or if we can just choose one and be done with it.

sweemer commented:
I fixed a few problems with the `TimeSeries` class in order to resolve a compilation issue with `test-suite/timeseries.cpp` after replacing `boost::unordered_map` with `std::unordered_map`.

- The previous code assumes that the container's `const_iterator::iterator_category` will be the same or derived from `std::bidirectional_iterator_tag` if and only if the container defines a `const_reverse_iterator` type, but this is not necessarily true. I believe that the confusion may have arisen from [cppreference.com](https://en.cppreference.com/w/cpp/container/unordered_map) saying that `std::unordered_map::const_iterator` is of type `LegacyForwardIterator`, but according to the actual [standard](http://eel.is/c++draft/unord.map), the type of `std::unordered_map::const_iterator` is implementation defined.
- The comments in `timeseries.hpp` say that "The following class makes compilation fail for the code that calls rbegin or rend with a container that does not support reverse iterators," but that's not how `TimeSeries` was actually implemented. Instead, compilation failed when `TimeSeries` was instantiated for any unordered container, even if the `rbegin` or `rend` functions were never called. I fixed the SFINAE to make it actually do what the comments claim (assuming that this is indeed still the desired functionality).
- The commented-out code in `test-suite/timeseries.cpp` called a private function so I have replaced that with a lambda. I verified that the commented-out code does in fact not compile after my changes.

Sorry for all the CI failure spam today. I should have waited a bit longer to submit the pull request, but some of these issues only popped up on platforms that I don't regularly compile on, and it's taking a bit longer to fix than I expected.

lballabio commented:
Don't worry about failure spam — I only get notifications for builds I trigger myself.

I guess what I was trying to say is, don't invest too much effort into making this work?  Unless it's your idea of fun, of course 😄 

I'll be happy if it works as advertised, but also if it keeps blowing up for some containers as it does now.  Apart from that test (that we can leave on boost, or rip out), I'm not sure if anybody uses anything but the default anyway.

sweemer commented:
I reverted the change to `test-suite/timeseries.cpp` and `ql/timeseries.hpp`. The checks are all passing now and are ready to merge.

If we rip out the test for `TimeSeries<int, boost::unordered_map<Date, int>>` then we can also get rid of all the `reverse` structs in `ql/timeseries.hpp`. As you say, it's hard to imagine anybody using an unordered container for time series both from a conceptual and a performance perspective. Of course it would be good to first deprecate `TimeSeries` used with unordered containers but I'm not sure of a good way to do this besides implementing partial specializations for all of the common unordered container types and deprecating each of those. Can you think of a better way? Or we can go straight to removal of `reverse` if you're fine with that approach.

lballabio commented:
I'm not sure I got the last comment — I wouldn't remove reverse iterators from TimeSeries so that it can be instantiated with `std::unordered_map`; reverse iterators are more useful than the latter.

lballabio commented on cmake/Platform.cmake:
```diff
@@ -18,8 +18,11 @@ if (MSVC)
     # /wd4267
     # Suppress warnings: assignment of 64-bit value to 32-bit QuantLib::Integer (x64)
 
+    # /wd4819
+    # Suppress warnings: The file contains a character that cannot be represented in the current code page
+
```
Where was the warning emitted?  I don't see it in the Appveyor build.

sweemer replied:
This warning is emitted on my Windows PC with active code page 949.

lballabio commented on ql/patterns/observable.hpp:
```diff
@@ -110,7 +111,7 @@ namespace QuantLib {
 
         // observer interface
-        std::pair<iterator, bool>
+        std::pair<boost::unordered_set<ext::shared_ptr<Observable>>::iterator, bool>
```
Deprecating `set_type` doesn't help if we have `boost::` exposed in the interface explicitly.

I would deprecate `set_type` (and make it private later) but leave the `iterator` typedef alone (and public) so it can be used in this signature.

sweemer replied:
Done

sweemer commented:
> I'm not sure I got the last comment — I wouldn't remove reverse iterators from TimeSeries so that it can be instantiated with `std::unordered_map`; reverse iterators are more useful than the latter.

Right, I'm not suggesting that we remove reverse iterators from `TimeSeries`. I'm suggesting that we remove the `reverse` class used to delegate the `rend` and `rbegin` methods to the underlying container type. Since the test fails to compile when replacing `boost::unordered_map` with `std::unordered_map` anyway, I figure it's better to explicitly not support unordered containers. The reverse iterators will still work for ordered containers of course - just not for unordered ones.

lballabio commented:
I'll merge and leave the time series alone for now.  Thanks!
